### DISCUSSION TITLE
Should I use MatchedValueBC or PenaltyInterfaceDiffusion to enforce solution continuity at an interface?

### DISCUSSION THREAD

**Discussion by lindsayad** [idaholab/moose#18799] [Q&A General] [answered]
The former enforces continuity strongly while the latter imposes continuity in a weak sense using a penalty method. Which one is better?

**Comment by lindsayad** [comment, marked as answer]
The answer is that it depends on what you're after. `MatchedValueBC` may yield a better conditioned system because it does not introduce any penalty parameters. However, this may not always be the case. For instance if you are solving a diffusivity problem, and you impose a penalty equivalent to the diffiusivity, then the penalty method will actually be better conditioned than the strongly enforced method. However, you may not be satisfied with the jump between the solution value on either side of the face.

What is unequivocal is that the penalty method (given wise choice of the penalty parameter) preserves optimal convergence of the finite element method while the strong method of enforcing continuity degrades the convergence by order 1. So if using first order Lagrange variables, the expected approximation error of the method is O(h^2). This is indeed the result when using the penalty method. However, in this case the strong enforcement method would yield error O(h^1). For more discussion please see the [old MOOSE forum](https://groups.google.com/g/moose-users/c/fi68gQRZl9g/m/nwSv-EDfBgAJ)

**Comment by jiangwen84** [comment]
@lindsayad  I was bit puzzled about the convergence rate you gave for strong enforcement v.s penalty method. Is it specific to the interfacial conditions? Or it is also true for boundary conditions?

For boundary condition, my experience is that the strong enforcement, i.e. `DirichletBC`, gives optimal rate while  `PenaltyDirichletBC` typically does not provide optimal rate convergence.  The penalty method requires theoretically very large number to give accurate result, and when the penalty becomes large, the optimal rate can not be obtained.

**Reply by lindsayad** [reply]
If you read that whole thread that I linked to, then I think you'll see that I was quite surprised as well! An analogy I can think of is that if we apply mechanical contact constraints in a point-wise way like we do for `NodeFaceConstraints`, then you fail the patch test and the method is not variationally consistent. However, if you apply the constraint in a weak/integral/weighted way then the variational consistency of the method is restored. @YaqiWang was the one who initially questioned the strong enforcement method and suggested that I look into the penalty method. I'm thankful that he did!

**Reply by lindsayad** [reply]
Instead of penalty we could also do weak enforcement with lagrange multipliers as well. Then you would get exact constraint enforcement (at the cost of making your system saddle point)

**Reply by jiangwen84** [reply]
I agreed that the `MatchedValueBC` is not a variational consistent approach, just like the NodeFace Contact which is more like a collocation method. For `MatchedValueBC`, I do not think it can be called as strong enforcement. It has its own residual that needs to be minimized by the solver.  Unlike the Dirichlet boundary condition, we can choose a finite element space which exactly satisfies the BC. For that we say it is a strong enforcement. 

For penalty, I am not sure if it is always optimal rate of convergence. It can be super or sub. Like you said, it is highly depends on the choice of the penalty parameter, which can be mesh dependent and problem dependent. 

For stabilized Lagrange Multiplier method, I believe it always has optimal rate of convergence.

**Reply by lindsayad** [reply]
> For `MatchedValueBC`, I do not think it can be called as strong enforcement. It has its own residual that needs to be minimized by the solver.

It is implemented as a `NodalBC` though. It overwrites any existing residual information for the degree of freedom on the node and plugs the residual of the `NodalBC` in. The only way the residual can be zero is if if the `NodalBC` residual is satisfied. In addiition, no information from the boundary faces is considered, e.g. no integration is performed. I would consider that strong enforcement.

**Reply by lindsayad** [reply]
> For penalty, I am not sure if it is always optimal rate of convergence. It can be super or sub. Like you said, it is highly depends on the choice of the penalty parameter, which can be mesh dependent and problem dependent.

This statement is true. At least I have observed sub-convergence rates with too small a choice of penalty. I have not yet observed super-convergence rates. I will say that just last night @fdkong tried out both the matched and penalty approaches and his error was significantly less with the penalty approach which is consistent with the plots from the MOOSE users thread I linked to.

> For stabilized Lagrange Multiplier method, I believe it always has optimal rate of convergence.

Yes, I think it would be great to add an object that does this with LMs

**Reply by jiangwen84** [reply]
If you look at the residual defined by `MatchedValueBC`, i.e., `_u[_qp] - _v[_qp]`, it is a penalty-like enforcement with penalty number of 1. I think if you user a higher penalty number there, you will get better results. 

In my opinion, those two methods are both penalty enforcement, with one enforces on the node, the other on the face/side.

**Reply by lindsayad** [reply]
No it is exact. `MatchedValueBC` is a `NodalBC`. `NodalBC` overwrites (e.g. it calls `NumericVector::set`) its residual and its Jacobian row instead of adding to it.

**Reply by jiangwen84** [reply]
I might not understand how it works well. Here is what I thought.

The residual cannot not be exactly zero. It will be minimized to meet the tolerance, e.g. L2 norm of the residual vector. So my understanding is that the `_u` won't be exactly equal to `_v`. If you add a larger penalty number to the residual term in `MatchedValueBC`, e.g. `1e6 * (_u[_qp] - _v[_qp])`, the contribution of this term to the overall residual vector will become larger. With a same tolerance, the `_u` will get closer to `_v`.

**Reply by lindsayad** [reply]
By that logic we should also add penalty parameters to Lagrange Multiplier residuals

**Reply by lindsayad** [reply]
So yes you are right ... if you're registering your problem as converged when the L2 norm of the residual vector is, say, 1e-8, then you are requiring that u and v are at most 1e-4 apart. If you add a penalty parameter, then you will force them to be even closer than that. My suspicion is that even if you add a penalty parameter to `MatchedValueBC`, no matter what value you have for it, you will never get the optimal convergence rate for the same reason that node-face constraints do not yield optimal convergence. But I could be wrong. @fdkong would you want to try adding a penalty parameter to `MatchedValueBC` and see if you get better errors, both better constant and slope?

**Reply by hugary1995** [reply]
I agree with Wen. I also want to point out that the tolerance argument is also problem dependent and solve scheme dependent. 

For a generally nonlinear problem, with a nonlinear Dirichlet BC like `MatchedValueBC`, depending on the nonlinearity of the problem, the dofs of the BC may converge before or after other dofs of the system. 

For example, if `v` is constant, then the matched value BC takes exactly one nonlinear iteration to converge, but the overall problem may need several iterations, and the BC will be exact. However, if `v` depends on other nonlinear variables in the system, then the tolerate issue will arise.

**Reply by lindsayad** [reply]
show me the convergence study and I'll be happy to agree!

**Reply by hugary1995** [reply]
Oh wait were you talking about mesh convergence? I was talking about convergence as in nonlinear iterations.

**Reply by lindsayad** [reply]
All my arguments have been about mesh convergence. I assumed that @jiangwen84 was talking about mesh convergence as well but looking back it looks like it's possible he was also talking about solve/nonlinear convergence?

**Reply by jiangwen84** [reply]
I was talking about the mesh convergence. I am not very concerned about the nonlinear convergence as long as the Jacobian is reasonably accurate.
